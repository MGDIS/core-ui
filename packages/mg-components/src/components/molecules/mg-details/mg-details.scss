@import '../../../styles/a11y.scss';

.mg-details {
  summary {
    display: flex;
    align-items: baseline;
    cursor: pointer;
  }

  &__toggle {
    margin-left: 1.5rem;
    flex-shrink: 0;
    display: flex;
    align-items: center;
    align-self: flex-start;
    gap: 0.6rem;
    min-height: calc(var(--font-size) * var(--line-height));
  }

  &__details {
    margin-top: var(--mg-details-spacing);
  }
}

@media (width < 70rem) {
  .mg-details__toggle > span {
    @include sr-only();
  }
}
