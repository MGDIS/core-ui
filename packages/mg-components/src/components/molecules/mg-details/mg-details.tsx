import { Component, Event, EventEmitter, h, Prop, Watch } from '@stencil/core';
import { isValidString } from '@mgdis/stencil-helpers';

@Component({
  tag: 'mg-details',
  styleUrl: '../../../../node_modules/@mgdis/styles/dist/components/mg-details.css',
  shadow: true,
})
export class MgDetails {
  /************
   * Internal *
   ************/

  // HTML selector
  private details: HTMLDetailsElement;

  /**************
   * Decorators *
   **************/

  /**
   * Displayed title when details are closed
   */
  @Prop() toggleClosed!: string;

  /**
   * Displayed title when details are opened
   */
  @Prop() toggleOpened!: string;
  @Watch('toggleClosed')
  @Watch('toggleOpened')
  validateTitles(newValue: string): void {
    if (!isValidString(newValue)) {
      throw new Error('<mg-details> prop "toggleClosed" and "toggleOpened" must be defined.');
    }
  }

  /**
   * Hide summary element
   */
  @Prop() hideSummary = false;

  /**
   * Define if details are diplayed
   */
  @Prop({ mutable: true }) expanded = false;
  @Watch('expanded')
  handleExpanded(newValue: boolean): void {
    this.expandedChange.emit(newValue);
  }

  /**
   * Emmited event when expanded change
   */
  @Event({ eventName: 'expanded-change' }) expandedChange: EventEmitter<HTMLMgDetailsElement['expanded']>;

  /**
   * Handle details toggle
   */
  private handleToggle = (): void => {
    this.expanded = this.details.open;
  };

  /*************
   * Lifecycle *
   *************/

  /**
   * Check if component props are well configured on init
   */
  componentWillLoad(): void {
    this.validateTitles(this.toggleClosed);
    this.validateTitles(this.toggleOpened);
  }

  /**
   * Render
   * @returns HTML Element
   */
  render(): HTMLElement {
    return (
      <details class="mg-c-details" onToggle={this.handleToggle} open={this.expanded} ref={(el: HTMLDetailsElement) => (this.details = el)}>
        <summary>
          <slot name="summary"></slot>
          <span class="mg-c-details__toggle">
            <mg-icon icon="chevron-up" size="small" class={{ 'mg-c-details__toggle-icon': true, 'mg-c-details__toggle-icon--reverse': !this.expanded }}></mg-icon>
            <span class={{ 'mg-u-visually-hidden': this.hideSummary }}>{this.expanded ? this.toggleOpened : this.toggleClosed}</span>
          </span>
        </summary>
        <div class="mg-c-details__details">
          <slot name="details"></slot>
        </div>
      </details>
    );
  }
}
