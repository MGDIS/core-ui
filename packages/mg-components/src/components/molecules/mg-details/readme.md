

<!-- Auto Generated Below -->


## Properties

| Property                    | Attribute       | Description                             | Type      | Default     |
| --------------------------- | --------------- | --------------------------------------- | --------- | ----------- |
| `expanded`                  | `expanded`      | Define if details are diplayed          | `boolean` | `false`     |
| `hideSummary`               | `hide-summary`  | Hide summary element                    | `boolean` | `false`     |
| `toggleClosed` _(required)_ | `toggle-closed` | Displayed title when details are closed | `string`  | `undefined` |
| `toggleOpened` _(required)_ | `toggle-opened` | Displayed title when details are opened | `string`  | `undefined` |


## Events

| Event             | Description                        | Type                   |
| ----------------- | ---------------------------------- | ---------------------- |
| `expanded-change` | Emmited event when expanded change | `CustomEvent<boolean>` |


## Slots

| Slot        | Description     |
| ----------- | --------------- |
| `"details"` | Details content |
| `"summary"` | Summary content |


## CSS Custom Properties

| Name                     | Description                                                                              |
| ------------------------ | ---------------------------------------------------------------------------------------- |
| `--mg-c-details-spacing` | Defines the spacing between the summary and the details. Default value: `--mg-b-size-8`. |


## Dependencies

### Depends on

- [mg-icon](../../atoms/mg-icon)

### Graph
```mermaid
graph TD;
  mg-details --> mg-icon
  style mg-details fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
