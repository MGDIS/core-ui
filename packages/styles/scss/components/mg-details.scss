@import '../utils/a11y';

.mg-details {
  summary {
    display: flex;
    align-items: baseline;
    cursor: pointer;
  }

  &__toggle {
    margin-left: 1.5rem;
    flex-shrink: 0;
    display: flex;
    align-items: center;
    align-self: flex-start;
    gap: 0.6rem;
    min-height: calc(var(--font-size) * var(--line-height));
    &-icon.mg-details__toggle-icon--reverse {
      transform: rotate(180deg);
    }
  }

  &__details {
    margin-top: var(--mg-details-spacing);
  }
}

/* hide toggle button when < 700px */
@media (width < 43.75rem) {
  .mg-details__toggle > span {
    @include visually-hidden();
  }
}
