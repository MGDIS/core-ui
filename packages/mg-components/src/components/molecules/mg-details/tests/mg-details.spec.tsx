import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { MgDetails } from '../mg-details';

const getPage = args =>
  newSpecPage({
    components: [MgDetails],
    template: () => (
      <mg-details {...args}>
        {args.slotSummary ? (
          args.slotSummary
        ) : (
          <span slot="summary">Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.</span>
        )}
        <p slot="details">
          Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse
          cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
        </p>
      </mg-details>
    ),
  });

describe('mg-details', () => {
  test.each([
    { toggleClosed: 'Show details', toggleOpened: 'Hide details' },
    { toggleClosed: 'Show details', toggleOpened: 'Hide details', expanded: true },
    { toggleClosed: 'Show details', toggleOpened: 'Hide details', hideSummary: true },
  ])('with args %s', async args => {
    const { root } = await getPage(args);
    expect(root).toMatchSnapshot();
  });

  test.each([
    {},
    { toggleClosed: 'Show details' },
    { toggleOpened: 'Hide details' },
    { toggleClosed: ' ', toggleOpened: 'Hide details' },
    { toggleClosed: 'Show details', toggleOpened: ' ' },
  ])('Should throw error when missing toggle messages: %s', async args => {
    expect.assertions(1);
    try {
      await getPage(args);
    } catch (err) {
      expect(err.message).toEqual(
        `<mg-details> prop "toggleClosed" and "toggleOpened" is required and must be a valid string. Passed value: "toggleClosed='${args.toggleClosed}'" and "toggleOpened='${args.toggleOpened}'".`,
      );
    }
  });

  test('Should toggle details on click', async () => {
    const page = await getPage({ toggleClosed: 'Show details', toggleOpened: 'Hide details' });

    const element = page.doc.querySelector('mg-details');
    const summary = element.shadowRoot.querySelector('summary');

    jest.spyOn(page.rootInstance.expandedChange, 'emit');

    summary.dispatchEvent(new CustomEvent('toggle', { bubbles: true }));
    await page.waitForChanges();

    expect(page.rootInstance.expandedChange.emit).toHaveBeenCalledTimes(1);
  });

  test.each([
    <span slot="summary">
      <mg-button>button</mg-button>
    </span>,
    <mg-button slot="summary">button</mg-button>,
  ])('Should prevent toggle details toggled on mg-button click %s', async slotSummary => {
    const page = await getPage({ toggleClosed: 'Show details', toggleOpened: 'Hide details', slotSummary });

    const element = page.doc.querySelector('mg-details');
    const mgButton = element.querySelector('mg-button');

    jest.spyOn(page.rootInstance.expandedChange, 'emit');

    mgButton.dispatchEvent(new CustomEvent('click', { bubbles: true }));
    await page.waitForChanges();

    expect(page.rootInstance.expandedChange.emit).toHaveBeenCalledTimes(0);
    expect(page.rootInstance.expanded).toEqual(false);
  });
});
