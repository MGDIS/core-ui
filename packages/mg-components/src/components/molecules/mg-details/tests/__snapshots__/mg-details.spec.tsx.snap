// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mg-details with args {
  toggleClosed: 'Show details',
  toggleOpened: 'Hide details',
  expanded: true
} 1`] = `
<mg-details>
  <template shadowrootmode="open">
    <details class="mg-c-details" open="">
      <summary>
        <slot name="summary"></slot>
        <span class="mg-c-details__toggle">
          <mg-icon class="mg-c-details__toggle-icon" icon="chevron-up"></mg-icon>
          <span>
            Hide details
          </span>
        </span>
      </summary>
      <div class="mg-c-details__details">
        <slot name="details"></slot>
      </div>
    </details>
  </template>
  <span slot="summary">
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
  </span>
  <p slot="details">
    Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
  </p>
</mg-details>
`;

exports[`mg-details with args {
  toggleClosed: 'Show details',
  toggleOpened: 'Hide details',
  hideSummary: true
} 1`] = `
<mg-details>
  <template shadowrootmode="open">
    <details class="mg-c-details">
      <summary>
        <slot name="summary"></slot>
        <span class="mg-c-details__toggle">
          <mg-icon class="mg-c-details__toggle-icon mg-c-details__toggle-icon--reverse" icon="chevron-up"></mg-icon>
          <span class="mg-u-visually-hidden">
            Show details
          </span>
        </span>
      </summary>
      <div class="mg-c-details__details">
        <slot name="details"></slot>
      </div>
    </details>
  </template>
  <span slot="summary">
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
  </span>
  <p slot="details">
    Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
  </p>
</mg-details>
`;

exports[`mg-details with args { toggleClosed: 'Show details', toggleOpened: 'Hide details' } 1`] = `
<mg-details>
  <template shadowrootmode="open">
    <details class="mg-c-details">
      <summary>
        <slot name="summary"></slot>
        <span class="mg-c-details__toggle">
          <mg-icon class="mg-c-details__toggle-icon mg-c-details__toggle-icon--reverse" icon="chevron-up"></mg-icon>
          <span>
            Show details
          </span>
        </span>
      </summary>
      <div class="mg-c-details__details">
        <slot name="details"></slot>
      </div>
    </details>
  </template>
  <span slot="summary">
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
  </span>
  <p slot="details">
    Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
  </p>
</mg-details>
`;

exports[`mg-details with empty summary slot 1`] = `
<mg-details slotsummary="">
  <template shadowrootmode="open">
    <details class="mg-c-details mg-c-details--empty-summary">
      <summary>
        <slot name="summary"></slot>
        <span class="mg-c-details__toggle">
          <mg-icon class="mg-c-details__toggle-icon mg-c-details__toggle-icon--reverse" icon="chevron-up"></mg-icon>
          <span>
            Show details
          </span>
        </span>
      </summary>
      <div class="mg-c-details__details">
        <slot name="details"></slot>
      </div>
    </details>
  </template>
  <p slot="details">
    Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
  </p>
</mg-details>
`;

exports[`mg-details with empty summary slot 2`] = `
<mg-details>
  <template shadowrootmode="open">
    <details class="mg-c-details mg-c-details--empty-summary">
      <summary>
        <slot name="summary"></slot>
        <span class="mg-c-details__toggle">
          <mg-icon class="mg-c-details__toggle-icon mg-c-details__toggle-icon--reverse" icon="chevron-up"></mg-icon>
          <span>
            Show details
          </span>
        </span>
      </summary>
      <div class="mg-c-details__details">
        <slot name="details"></slot>
      </div>
    </details>
  </template>
  <span slot="summary"></span>
  <p slot="details">
    Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
  </p>
</mg-details>
`;

exports[`mg-details with empty summary slot 3`] = `
<mg-details>
  <template shadowrootmode="open">
    <details class="mg-c-details mg-c-details--empty-summary">
      <summary>
        <slot name="summary"></slot>
        <span class="mg-c-details__toggle">
          <mg-icon class="mg-c-details__toggle-icon mg-c-details__toggle-icon--reverse" icon="chevron-up"></mg-icon>
          <span>
            Show details
          </span>
        </span>
      </summary>
      <div class="mg-c-details__details">
        <slot name="details"></slot>
      </div>
    </details>
  </template>
  <span slot="summary"></span>
  <p slot="details">
    Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
  </p>
</mg-details>
`;
